internal/errors.js

```diff
@@ -13,133 +13,41 @@
-        // Reset the limit and setting the name property.
-        Error.stackTraceLimit = limit;
-      }
-      const message = getMessage(key, args, this);
-      Object.defineProperty(this, 'message', {
-        value: message,
-        enumerable: false,
-        writable: true,
-        configurable: true
-      });
-      addCodeToName(this, super.name, key);
-    }
-
-    get code() {
-      return key;
-    }
-
-    set code(value) {
-      defineProperty(this, 'code', {
-        configurable: true,
-        enumerable: true,
-        value,
-        writable: true
-      });
-    }
-
-    toString() {
-      return `${this.name} [${key}]: ${this.message}`;
-    }
-  };
-}
-
-function addCodeToName(err, name, code) {
-  // Set the stack
-  if (excludedStackFn !== undefined && Error.captureStackTrace) {
-    // eslint-disable-next-line no-restricted-syntax
-    Error.captureStackTrace(err, excludedStackFn);
+function createErrorType(code, message, Base) {
+  if (!Base) {
+    Base = Error
   }
-  // Add the error code to the name to include it in the stack trace.
-  err.name = `${name} [${code}]`;
-  // Access the stack to generate the error message including the error code
-  // from the name.
-  err.stack;
-  // Reset the name to the actual name.
-  if (name === 'SystemError') {
-    defineProperty(err, 'name', {
-      value: name,
-      enumerable: false,
-      writable: true,
-      configurable: true
-    });
-  } else {
-    delete err.name;
-  }
-}
-
-// Utility function for registering the error codes. Only used here. Exported
-// *only* to allow for testing.
-function E(sym, val, def, ...otherClasses) {
-  messages.set(sym, val);
-  def = makeNodeErrorWithCode(def, sym);
 
-  if (otherClasses.length !== 0) {
-    otherClasses.forEach((clazz) => {
-      def[clazz.name] = makeNodeErrorWithCode(clazz, sym);
-    });
+  function getMessage (arg1, arg2, arg3) {
+    if (typeof message === 'string') {
+      return message
+    } else {
+      return message(arg1, arg2, arg3)
+    }
   }
-  codes[sym] = def;
-}
 
-function getMessage(key, args, self) {
-  if (assert === undefined) assert = require('../assert');
-  const msg = messages.get(key);
-
-  if (typeof msg === 'function') {
-    assert(
-      msg.length <= args.length, // Default options do not count.
-      `Code: ${key}; The provided arguments length (${args.length}) does not ` +
-        `match the required ones (${msg.length}).`
-    );
-    return msg.apply(self, args);
+  class NodeError extends Base {
+    constructor (arg1, arg2, arg3) {
+      super(getMessage(arg1, arg2, arg3));
+    }
   }
 
-  const expectedLength = (msg.match(/%[dfijoOs]/g) || []).length;
-  assert(
-    expectedLength === args.length,
-    `Code: ${key}; The provided arguments length (${args.length}) does not ` +
-      `match the required ones (${expectedLength}).`
-  );
-  if (args.length === 0)
-    return msg;
+  NodeError.prototype.name = Base.name;
+  NodeError.prototype.code = code;
```

Comment:
This is probably the issue. Check if it's sufficient to assign the code property inside of the constructor instead as in:

```js
constructor (arg1, arg2, arg3) {
  super(getMessage(arg1, arg2, arg3));
  this.code = code;
}
```

Reply:
Just in time!

I'd actually just fixed it locally with:

```js
function NodeError(arg1, arg2, arg3) {
  const baseError = new Base(getMessage(arg1, arg2, arg3));
  baseError.code = code;
  return baseError;
}
```

Although your solution also works and looks much less hacky.

With your solution should I also remove:

```js
NodeError.prototype.name = Base.name;
NodeError.prototype.code = code;
```

I leave them in?

Reply:
And why do we need to set `code` inside the constructor but not `name`?

Reply:
Oh, because `name` is already inherited from `Base` but `code` isn't?

Reply:
> Oh, because name is already inherited from Base but code isn't?

The name property is somewhat special on errors. The base name is used as long as the actual class has no own `name` property (in which case it would use that instead).

It also seems the `code` property is not properly transpiled.

It should be fine to remove both lines.

Reply:
Finally, why we don't receive and don't set `name`?

Reply:
Finally, why we don't receive and don't set `name`?

Reply:
Finally, why we don't receive and don't set `name`?

Reply:
@kuraga the constructor has the name. The error itself also has a name but for Node.js errors it's not set on the instance, so the inherited name is used.